Zend/zend_operators.c

```diff
@@ -1287,6 +1287,21 @@ ZEND_API zend_result ZEND_FASTCALL mul_function(zval *result, zval *op1, zval *o
 }
 /* }}} */
 
+static void ZEND_COLD zend_power_base_0_exponent_lt_eq_0_error(void)
+{
+	zend_error(E_DEPRECATED, "Power of base 0 and negative exponent is deprecated");
+}
+
+static zend_result safe_pow(double *result, double base, double exponent)
+{
+	if (base == 0.0 && exponent < 0.0) {
+		zend_power_base_0_exponent_lt_eq_0_error();
```

Comment:
This should likely check for `EG(exception)` and return `FAILURE` in that case (for throwing error handlers). Usages of `safe_pow` should propagate this return value.

Reply:
I'm not sure what I should change here. The changes should not throw an exception but the deprecation note. In PHP 9.0 it should throw an exception, but this PR is aimed at 8.4. The same note applies to your second comment.

Reply:
Warnings can be promoted from error handlers to exceptions. In these cases, often, the program flow must be adjusted in some way. Anyway, I looked at the way `pow_function_base()`s return value is used, and apparently it indicates not whether the operation was successful, but whether the two operand types support the `pow` operation. So a change is probably not necessary here.